junit-platform-engine/src/main/java/org/junit/platform/engine/support/discovery/ResourceContainerSelectorResolver.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2015-2024 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.engine.support.discovery;
+
+import static java.util.stream.Collectors.toSet;
+import static org.junit.platform.commons.support.ReflectionSupport.findAllResourcesInClasspathRoot;
+import static org.junit.platform.commons.support.ReflectionSupport.findAllResourcesInPackage;
+import static org.junit.platform.commons.util.ReflectionUtils.findAllResourcesInModule;
+import static org.junit.platform.engine.support.discovery.SelectorResolver.Resolution.selectors;
+import static org.junit.platform.engine.support.discovery.SelectorResolver.Resolution.unresolved;
+
+import java.util.List;
+import java.util.function.Predicate;
+
+import org.junit.platform.commons.support.Resource;
+import org.junit.platform.engine.discovery.ClasspathRootSelector;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.engine.discovery.ModuleSelector;
+import org.junit.platform.engine.discovery.PackageSelector;
+
+/**
+ * @since 1.11
+ */
+class ResourceContainerSelectorResolver implements SelectorResolver {
+	private static final char CLASSPATH_RESOURCE_PATH_SEPARATOR = '/';
+	private static final char PACKAGE_SEPARATOR_CHAR = '.';
+	public static final String DEFAULT_PACKAGE_NAME = "";
+	private final Predicate<Resource> resourceFilter;
+
+	ResourceContainerSelectorResolver(Predicate<Resource> resourceFilter, Predicate<String> resourcePackageFilter) {
+		Predicate<Resource> packageFilter = adaptPackageFilter(resourcePackageFilter);
+		this.resourceFilter = packageFilter.and(resourceFilter);
+	}
+
+	/**
+	 * A package filter is written to test {@code .} separated package names.
+	 * Resources however have {@code /} separated paths. By rewriting the path
+	 * of the resource into a package name, we can make the package filter work.
+	 */
+	private static Predicate<Resource> adaptPackageFilter(Predicate<String> packageFilter) {
+		return resource -> packageFilter.test(packageName(resource.getName()));
+	}
+
+	private static String packageName(String classpathResourceName) {
+		int lastIndexOf = classpathResourceName.lastIndexOf(CLASSPATH_RESOURCE_PATH_SEPARATOR);
+		if (lastIndexOf < 0) {
+			return DEFAULT_PACKAGE_NAME;
+		}
+		// classpath resource names do not start with /
+		String resourcePackagePath = classpathResourceName.substring(0, lastIndexOf);
+		return resourcePackagePath.replace(CLASSPATH_RESOURCE_PATH_SEPARATOR, PACKAGE_SEPARATOR_CHAR);
```

Comment:
I'm a bit worried about doing this for every candidate resource but apart from a limited-size cache I don't have a better idea.

Reply:
We could short circuit this if the predicate accepts all packages.

I'll see if I can performance tests this.

Reply:
Okay. Performance impact is pretty negligible.

Profiling `ResourceContainerSelectorResolver.resolve` with 160k resource files in a small directory structure, takes  ~4600ms, most of which is spend in either `ClassLoader.getResource` or `FileTreeWalker.next` The time spend in `packageName` is ~10ms. Adding caching increases that to ~20ms. Were you perhaps thinking of `String.replaceAll`? 

![image](https://github.com/junit-team/junit5/assets/6946919/1acd351a-3046-4516-afe4-4845885e732c)

Reply:
Thanks for checking! Looks like my worry was unfounded.